src/kernel-install/60-ukify.install.in

```diff
@@ -0,0 +1,225 @@
+    if p.exists():
+        return p
+    return None
+
+
+def kernel_cmdline_base() -> typing.Optional[typing.List[str]]:
+    if root := os.getenv('KERNEL_INSTALL_CONF_ROOT'):
+        return Path(root).joinpath('cmdline').read_text().split()
+
+    for cmdline in ('/etc/kernel/cmdline',
+                    '/usr/lib/kernel/cmdline'):
+        try:
+            return Path(cmdline).read_text().split()
+        except FileNotFoundError:
+            continue
+
+    options = Path(cmdline).read_text().split()
+    return [opt for opt in options
+            if not opt.startswith(('BOOT_IMAGE=', 'initrd='))]
+
+
+def kernel_cmdline(opts) -> str:
+    options = kernel_cmdline_base()
+
+    # If the boot entries are named after the machine ID, then suffix the kernel
+    # command line with the machine ID we use, so that the machine ID remains
+    # stable, even during factory reset, in the initrd (where the system's machine
+    # ID is not directly accessible yet), and if the root file system is volatile.
+    if (opts.entry_token == opts.machine_id and
+        not any(opt.startswith('systemd.machine_id=') for opt in options)):
+        options += [f'systemd.machine_id={opts.machine_id}']
+
+    # TODO: we unconditionally set the cmdline here, ignoring the setting in
+    #       the config file. Should we not do that?
+
+    # Prepend a space so that '@' does not get misinterpreted
+    return ' ' + ' '.join(options)
+
+
+def call_ukify(opts):
+    # Punish me harder.
+    # We want this:
+    #   ukify = importlib.machinery.SourceFileLoader('ukify', UKIFY).load_module()
+    # but it throws a DeprecationWarning.
+    # https://stackoverflow.com/questions/67631/how-can-i-import-a-module-dynamically-given-the-full-path
+    # https://github.com/python/cpython/issues/65635
+    # offer "explanations", but to actually load a python file without a .py extension,
+    # the "solution" is 4+ incomprehensible lines.
+    # The solution with runpy gives a dictionary, which isn't great, but will do.
+    ukify = runpy.run_path(UKIFY, run_name='ukify')
+
+    # Create "empty" namespace. We want to override just a few settings,
+    # so it doesn't make sense to duplicate all the fields. We use a hack
+    # to pre-populate the namespace like argparse would, all defaults.
+    # We need to specify the two mandatory arguments to not get an error.
+    opts2 = ukify['create_parser']().parse_args(('A','B'))
+
+    opts2.config = config_file_location()
+    opts2.uname = opts.kernel_version
+    opts2.linux = opts.kernel_image
+    opts2.initrd = opts.initrd
+    # Note that 'uki.efi' is the name required by 90-uki-copy.install.
+    opts2.output = opts.staging_area / 'uki.efi'
+
+    opts2.cmdline = kernel_cmdline(opts)
+    if BOOT_STUB:
+        opts2.stub = BOOT_STUB
+
+    # opts2.summary = True
+
+    ukify['apply_config'](opts2)
+    ukify['finalize_options'](opts2)
+    ukify['check_inputs'](opts2)
+    ukify['make_uki'](opts2)
```

Comment:
btw, should these functions in ukify.py carry some comment or so that clarifies they are entrypoints from this external script? just so that people don't forget to update everything when they change it.

Reply:
All functions that are public (no "_" prefix) are implicitly entry points… I think the CI should be enough: if one of those is ever renamed, we'll immediately get an error here. We distribute both files together, so we don't need cross-version compatibility.